File: Shared/ObjC/Config/SpreedlyConfigManager.m
//
//  SpreedlyConfigManager.m
//  SpreedlySDKExampleObjectiveC
//
//  Created by Vinay Naikade on 12/08/25.
//
//  This manager sets up Spreedly configuration and global theme
//  for all examples in the Objective-C sample app.
//

#import "SpreedlyConfigManager.h"
#import <SpreedlyCore/SpreedlyCore-Swift.h>
// #import <SpreedlyUI/SpreedlyUI-Swift.h>
#import "SignatureSecurityService.h"
#import "RetainPaymentMethodAPIClient.h"
#import "FetchPaymentMethodsAPIClient.h"
#import "PurchaseAPIClient.h"
#import "SignatureServerClient.h"

// Forward declaration for Swift classes
// @class SpreedlyThemeManagerObjC;
// @class SPLThemeConfig;

@interface SpreedlyConfigManager ()

@property (nonatomic, strong) NSString *environmentKey;
@property (nonatomic, strong) NSString *forterSiteId;
@property (nonatomic, strong) NSString *serverURL;
@property (nonatomic, strong) NSString *paymentMethodsBaseURL;
@property (nonatomic, strong) NSString *apiKey;

@end

@implementation SpreedlyConfigManager

static SpreedlyConfigManager *_shared = nil;

+ (SpreedlyConfigManager *)shared {
    return _shared;
}

+ (NSString *)infoPlistValueForKey:(NSString *)key {
    return [[NSBundle mainBundle] infoDictionary][key] ?: @"";
}

- (instancetype)init {
    if (self = [super init]) {
        self.environmentKey       = [SpreedlyConfigManager infoPlistValueForKey:@"SpreedlyEnvironmentKey"];
        self.forterSiteId         = [SpreedlyConfigManager infoPlistValueForKey:@"SpreedlyForterSiteId"];
        self.serverURL            = [SpreedlyConfigManager infoPlistValueForKey:@"SpreedlyServerURL"];
        self.paymentMethodsBaseURL = [SpreedlyConfigManager infoPlistValueForKey:@"SpreedlyBaseURL"];
        self.apiKey               = [SpreedlyConfigManager infoPlistValueForKey:@"SpreedlyApiKey"];
        
        [Spreedly initializeSDK];
    }
    return self;
}

+ (void)setup {
    _shared = [[SpreedlyConfigManager alloc] init];
    
    // // Set global theme for all SpreedlyUI components across all examples
    // SPLThemeConfig *globalTheme = [[SPLThemeConfig alloc] initWithPrimaryColorHex:@"#0077C8"
    //     secondaryColorHex:@"#AFB4B5"
    //     formBorderColorHex:@"#D9D9D9"
    //     formBackgroundColorHex:@"#FFFFFF"
    //     fieldBackgroundColorHex:@"#F8F9FA"
    //     fieldLabelColorHex:@"#6C757D"
    //     borderRadius:8.0];
    // [SpreedlyThemeManagerObjC setGlobalThemeWithConfig:globalTheme];
}

- (void)generateSignatureWithCompletion:(void (^)(BOOL success, NSError * _Nullable error))completion {
    // Create server security config
    ServerSecurityConfig *config = [[ServerSecurityConfig alloc] initWithServerURL:self.serverURL
                                                                             apiKey:self.apiKey.length > 0 ? self.apiKey : nil
                                                                     environmentKey:self.environmentKey];
    
    // Setup server-based security
    [SignatureSecurityService setupServerBasedSecurityWithConfig:config completion:^(SecuritySetupResult * _Nullable result) {
        if (!result.success) {
            if (completion) {
                completion(NO, result.error);
            }
            return;
        }
        
        if (!result.signatureParams) {
            NSError *signatureError = [NSError errorWithDomain:@"SpreedlyConfigManager"
                                                          code:0
                                                      userInfo:@{NSLocalizedDescriptionKey: @"Failed to generate signature parameters"}];
            if (completion) {
                completion(NO, signatureError);
            }
            return;
        }
        
        // Signature parameters generated (values not logged for security)
        // Setup Spreedly with config including forterSiteId
        SpreedlyConfig *spreedlyConfig = [[SpreedlyConfig alloc] initWithEnvironmentKey:self.environmentKey];
        spreedlyConfig.forterSiteId = self.forterSiteId;
        spreedlyConfig.certificateToken = result.signatureParams.certificateToken;
        spreedlyConfig.nonce = result.signatureParams.nonce;
        spreedlyConfig.signature = result.signatureParams.signature;
        spreedlyConfig.timestamp = [NSString stringWithFormat:@"%ld", (long)result.signatureParams.timestamp];
        
        [Spreedly setupWithConfig:spreedlyConfig];
        
        if (Spreedly.initializationError != nil) {
            NSError *blockError = [NSError errorWithDomain:@"SpreedlyConfigManager"
                                                      code:1
                                                  userInfo:@{NSLocalizedDescriptionKey:
                [NSString stringWithFormat:@"SDK blocked: %@", Spreedly.initializationError.message]}];
            if (completion) {
                completion(NO, blockError);
            }
            return;
        }
        
        if (completion) {
            completion(YES, nil);
        }
    }];
}

- (RetainPaymentMethodAPIClient *)createRetainPaymentMethodAPIClient {
    ServerConfig *config = [[ServerConfig alloc] initWithBaseURL:self.paymentMethodsBaseURL
                                                  timeoutInterval:30.0
                                                           apiKey:self.apiKey.length > 0 ? self.apiKey : nil];
    return [[RetainPaymentMethodAPIClient alloc] initWithConfig:config];
}

- (FetchPaymentMethodsAPIClient *)createFetchPaymentMethodsAPIClient {
    ServerConfig *config = [[ServerConfig alloc] initWithBaseURL:self.paymentMethodsBaseURL
                                                timeoutInterval:30.0
                                                         apiKey:self.apiKey.length > 0 ? self.apiKey : nil];
    return [[FetchPaymentMethodsAPIClient alloc] initWithConfig:config];
}

- (PurchaseAPIClient *)createPurchaseAPIClient {
    ServerConfig *config = [[ServerConfig alloc] initWithBaseURL:self.paymentMethodsBaseURL
                                                  timeoutInterval:30.0
                                                           apiKey:nil];
    return [[PurchaseAPIClient alloc] initWithConfig:config];
}

- (NSString *)stripePublishableKey {
    return [SpreedlyConfigManager infoPlistValueForKey:@"StripePublishableKey"];
}

@end 

File: Shared/Swift/Config/SpreedlyConfigManager.swift
//
//  SpreedlyConfigManager.swift
//  SpreedlySDKExample
//
//  Created by Vinay Naikade on 23/07/25.
//

import Foundation
import SpreedlyCore

class SpreedlyConfigManager {
    static var shared: SpreedlyConfigManager!

    private static func infoPlistValue(forKey key: String) -> String {
        Bundle.main.infoDictionary?[key] as? String ?? ""
    }
    
    private let environmentKey: String = infoPlistValue(forKey: "SpreedlyEnvironmentKey")
    private let forterSiteId: String = infoPlistValue(forKey: "SpreedlyForterSiteId")
    private let serverURL: String = infoPlistValue(forKey: "SpreedlyServerURL")
    private let baseURL: String = infoPlistValue(forKey: "SpreedlyBaseURL")
    let stripePublishableKey: String = infoPlistValue(forKey: "StripePublishableKey")
    private let apiKey: String = infoPlistValue(forKey: "SpreedlyApiKey")
    
    private init() {
        Spreedly.initializeSDK()
    }
    
    static func setup() {
        shared = SpreedlyConfigManager()
    }
    
    @discardableResult
    func generateSignature() async -> Result<Bool, Error> {
        let config = SignatureSecurityService.ServerSecurityConfig(
            serverURL: serverURL,
            apiKey: apiKey.isEmpty ? nil : apiKey,
            environmentKey: environmentKey
        )
        let result = await SignatureSecurityService.setupServerBasedSecurity(config: config)
        guard let signatureParams = result.signatureParams else {
            return .failure(result.error ?? NSError(domain: "SpreedlyConfigManager", code: 0, userInfo: [NSLocalizedDescriptionKey: "Failed to generate signature parameters"]))
        }
        let spreedlyConfig = SpreedlyConfig(
            environmentKey: environmentKey,
            forterSiteId: forterSiteId,
            certificateToken: signatureParams.certificateToken,
            nonce: signatureParams.nonce,
            signature: signatureParams.signature,
            timestamp: String(signatureParams.timestamp)
        )
        Spreedly.setup(config: spreedlyConfig)
        if let error = Spreedly.initializationError {
            return .failure(NSError(
                domain: "SpreedlyConfigManager",
                code: 1,
                userInfo: [NSLocalizedDescriptionKey: "SDK blocked: \(error.message)"]
            ))
        }
        return .success(true)
    }
    
    /// Creates a RetainPaymentMethodAPIClient with configured settings
    func createRetainPaymentMethodAPIClient() -> RetainPaymentMethodAPIClient {
        let config = RetainPaymentMethodAPIClient.ServerConfig(
            baseURL: baseURL,
            timeoutInterval: 30.0,
            apiKey: apiKey.isEmpty ? nil : apiKey
        )
        return RetainPaymentMethodAPIClient(config: config)
    }
    
    /// Creates a FetchPaymentMethodsAPIClient with configured settings
    func createFetchPaymentMethodsAPIClient() -> FetchPaymentMethodsAPIClient {
        let config = FetchPaymentMethodsAPIClient.ServerConfig(
            baseURL: baseURL,
            timeoutInterval: 30.0,
            apiKey: apiKey.isEmpty ? nil : apiKey
        )
        return FetchPaymentMethodsAPIClient(config: config)
    }
    
    /// Creates a PurchaseAPIClient with configured settings (Heroku base URL only; no gateway tokens).
    func createPurchaseAPIClient() -> PurchaseAPIClient {
        let config = PurchaseAPIClient.ServerConfig(
            baseURL: baseURL,
            timeoutInterval: 30.0
        )
        return PurchaseAPIClient(config: config)
    }
}
